vendor/github.com/robert-milan/go-object-interning/object_intern.go

```diff
@@ -0,0 +1,484 @@
+			b, err := shoco.Decompress(in)
+			return b, err
+		}
+	default:
+		oi.compress = func(in []byte) []byte {
+			return in
+		}
+		oi.decompress = func(in []byte) ([]byte, error) {
+			return in, nil
+		}
+	}
+
+	return oi
+}
+
+// CompressionFunc returns the current compression func used by the library
+func (oi *ObjectIntern) CompressionFunc() func(in []byte) []byte {
+	return oi.compress
+}
+
+// DecompressionFunc returns the current decompression func used by the library
+func (oi *ObjectIntern) DecompressionFunc() func(in []byte) ([]byte, error) {
+	return oi.decompress
+}
+
+// Compress returns a compressed version of in as a []byte
+// It is important to keep in mind that not all values can be compressed,
+// so this may at times return the original value
+func (oi *ObjectIntern) Compress(in []byte) []byte {
+	return oi.compress(in)
+}
+
+// Decompress returns a decompressed version of in as a []byte and nil on success.
+// On failure it returns an error
+func (oi *ObjectIntern) Decompress(in []byte) ([]byte, error) {
+	return oi.decompress(in)
+}
+
+// CompressSz returns a compressed version of in as a string
+// It is important to keep in mind that not all values can be compressed,
+// so this may at times return the original value
+func (oi *ObjectIntern) CompressSz(in string) string {
+	if oi.conf.CompressionType == NOCPRSN {
+		return in
+	}
+	return string(oi.compress([]byte(in)))
+}
+
+// DecompressSz returns a decompressed version of string as a string and nil upon success.
+// On failure it returns in and an error.
+func (oi *ObjectIntern) DecompressSz(in string) (string, error) {
+	if oi.conf.CompressionType == NOCPRSN {
+		return in, nil
+	}
+	b, err := oi.decompress([]byte(in))
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+// AddOrGet finds or adds and then returns a uintptr to an object and nil.
+// On failure it returns 0 and an error
+//
+// If the object is found in the store its reference count is increased by 1.
+// If the object is added to the store its reference count is set to 1.
+func (oi *ObjectIntern) AddOrGet(obj []byte) (uintptr, error) {
+	var addr gos.ObjAddr
+	var ok bool
+	var err error
+	var objSz string
+
+	// compress the object before searching for it
+	objComp := oi.compress(obj)
+	// if compression is turned off we don't want to work on the original
+	if oi.conf.CompressionType == NOCPRSN {
+		objComp := make([]byte, len(obj))
+		copy(objComp, obj)
+	}
+	objSz = string(objComp)
+
+	// acquire lock
+	oi.Lock()
+
+	// try to find the object in the index
+	addr, ok = oi.ObjIndex[objSz]
+	if ok {
+		// increment reference count by 1
+		(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
+		oi.Unlock()
+		return addr, nil
+	}
+
+	// The object is not in the index therefore it is not in the store.
+	// We need to set its initial reference count to 1 before adding it
+	objComp = append(objComp, []byte{0x1, 0x0, 0x0, 0x0}...)
```

Comment:
this looks like magic. we should probably document the "format" of the data (e.g. where is the refcount)